### DISCUSSION TITLE
How to use JBrowse CLI behind a firewall

### DISCUSSION THREAD

**Discussion by zhanxw** [GMOD/jbrowse-components#2539] [Q&A] [answered]
Thanks for developing JBrowse.

JBrowse CLI (create command) crashed due to an error (see below).
It seems that JBrowse CLI (or node-fetch) cannot recognize proxy settings.
May I suggest adding supports for proxy settings?

````
> jbrowse create jbrowse21
    FetchError: request to https://api.github.com/repos/GMOD/jbrowse-components/releases?page=1 failed, reason: read ECONNRESET
    Code: ECONNRESET
````

**Comment by cmdcolin** [comment, marked as answer]
Hi there

This is a good question
I see you got jbrowse CLI installed which is probably at least one hurdle on a firewall machine, but you can work around this issue by downloading e.g. https://github.com/GMOD/jbrowse-components/releases/download/v1.5.1/jbrowse-web-v1.5.1.zip manually and unzipping it instead of using jbrowse create. That is all jbrowse create does :)

Then you can use the other commands like add-assembly,add-track,text-index, etc. and it won't perform any network accesses generally

Edit: we now support fetching http proxy settings in v1.7.11+

**Reply by zhanxw** [reply]
Thanks for your answer. It solves my problem.

**Comment by AlbatrosHurleur** [comment]
Hi,

I'm facing the same issue with a proxy, and manually downloading and unzipping the latest release worked well as a workaround.

However, how would you mimic `jbrowse upgrade --clean myfolder` behind a proxy?

Thanks!

**Reply by cmdcolin** [reply]
hi there, I don't actually have any experience with proxies but I found this package which we could possibly add to the CLI https://www.npmjs.com/package/node-fetch-with-proxy 

if this was added, then possibly you can set e.g. export HTTP_PROXY=http://your.proxy:8888 and run jbrowse CLI like normal. does it seem like that environment variable would make sense with your use case?

**Reply by AlbatrosHurleur** [reply]
Hi Colin,

Thanks for your reply, yes it would make sense as HTTP_PROXY, HTTPS_PROXY and NO_PROXY variables are very often used for many applications when behind a proxy (that is also my case).

I'd be happy to try it and give a feedback if you add this support!

**Reply by cmdcolin** [reply]
If you wanted to try out a possible fix, you can try this

```
git clone https://github.com/GMOD/jbrowse-components
cd jbrowse-components
yarn
git checkout node_fetch_with_proxy # this is a branch using that library
cd products/jbrowse-cli
bin/run create newdir # possibly set the environment variables before this
```

**Reply by AlbatrosHurleur** [reply]
This is the output, I don't know JS but it seems there is one import that is not found:

```
$ git clone https://github.com/GMOD/jbrowse-components
Cloning into 'jbrowse-components'...
remote: Enumerating objects: 125553, done.
remote: Counting objects: 100% (3601/3601), done.
remote: Compressing objects: 100% (1487/1487), done.
remote: Total 125553 (delta 2353), reused 3173 (delta 2084), pack-reused 121952
Receiving objects: 100% (125553/125553), 163.53 MiB | 27.79 MiB/s, done.
Resolving deltas: 100% (89952/89952), done.
$ cd jbrowse-components
$ git checkout node_fetch_with_proxy # this is a branch using that library
Branch 'node_fetch_with_proxy' set up to track remote branch 'node_fetch_with_proxy' from 'origin'.
Switched to a new branch 'node_fetch_with_proxy'
$ cd products/jbrowse-cli
$ jbrowse create -l # jbrowse v1.7.10 - Not supposed to be working behind a proxy
    FetchError: request to https://api.github.com/repos/GMOD/jbrowse-components/releases?page=1 failed, reason: connect ETIMEDOUT 140.82.121.6:443
    Code: ETIMEDOUT
$ bin/run create jbrowse2
node:internal/modules/cjs/loader:936
  throw err;
  ^

Error: Cannot find module '@oclif/command'
Require stack:
- /home/sysadmin/jbrowse-components/products/jbrowse-cli/bin/run
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function.Module._load (node:internal/modules/cjs/loader:778:27)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/home/sysadmin/jbrowse-components/products/jbrowse-cli/bin/run:3:1)
    at Module._compile (node:internal/modules/cjs/loader:1105:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [ '/home/sysadmin/jbrowse-components/products/jbrowse-cli/bin/run' ]
}
```

**Reply by cmdcolin** [reply]
forgot to mention, must run `yarn` to install the dependencies, forgot to add that to my instructions. if you don't have yarn installed can say `npm install -g yarn` to install or `npx yarn` for one time usage

**Reply by AlbatrosHurleur** [reply]
Thanks, I installed yarn and the feature branch worked like a charm:

```
$ bin/run create -l
All JBrowse versions:
v1.7.10
v1.7.9
v1.7.8
v1.7.7
...
$ bin/run create jbrowse2
Fetching https://github.com/GMOD/jbrowse-components/releases/download/v1.7.10/jbrowse-web-v1.7.10.zip...
Unpacked https://github.com/GMOD/jbrowse-components/releases/download/v1.7.10/jbrowse-web-v1.7.10.zip at jbrowse2
```

If this get integrated in a release, it would be nice to also add an entry in the FAQ.

Thanks again for your help!

**Reply by cmdcolin** [reply]
excellent :) will add it in the next release!

**Reply by AlbatrosHurleur** [reply]
Great, thank you!

**Comment by cmdcolin** [comment]
support for e.g. HTTP_PROXY variables added in v1.7.11. thanks to all for the push!
